### PR TITLE
VR menus

### PR DESCRIPTION
**Pull request by jherico**
This CR brings the full application menu into the 3D scene.  Currently it can be activated by tapping the Alt key.  The state of the VR menu should reflect the state of the app menu at all times with the exception that things that have been hidden on the app menu still appear on the VR menu, but disabled.  This is a known bug.

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: 76ce34a74b3d4243d5a8d2b17c12f43e4e2422ef available at: https://gist.github.com/60ed53eab95ea3b520e0<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4700.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4700.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: ca16d37ce5fe70dd62cc24eafb5049a781bba73e available at: https://gist.github.com/60ed53eab95ea3b520e0<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4700.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4700.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 1beaf18d1baf096bec57aff78c70397fb76d1597 available at: https://gist.github.com/60ed53eab95ea3b520e0<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4700.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4700.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 59b8b8014865359e5b05adbb26f63d1b4f7591ec available at: https://gist.github.com/60ed53eab95ea3b520e0<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4700.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4700.dmg

**Comment by jherico** [comment]
test this please

**Comment by jherico** [comment]
test this please

**Comment by murillodigital** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Build console output for commit: 10b59350630553a500c912aa9d57f23c6a303787 available at: https://gist.github.com/60ed53eab95ea3b520e0<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4700.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4700.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 19ca543c2c9ab177e5343f88d8595307a00ddd23 available at: https://gist.github.com/60ed53eab95ea3b520e0<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4700.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4700.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: b0eccc40c27b0cb8a87af5186393ca663058f960 available at: https://gist.github.com/60ed53eab95ea3b520e0<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4700.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4700.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: d1d1830b2f811ae181b7055a82bab7878ef85afb available at: https://gist.github.com/60ed53eab95ea3b520e0<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4700.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4700.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: b4e9dbedbcc54c627c34ed8a4bdbc060f6a7a9b4 available at: https://gist.github.com/60ed53eab95ea3b520e0<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4700.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4700.dmg

File: interface/resources/qml/controls/CheckBox.qml [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+    style: CheckBoxStyle {
+        indicator: FontAwesome {
+            text: control.checked ? "\uf046" : "\uf096"
+        }
+        label: Text {
+            text: control.text
+        }
+    }
+    
+}
```
Comment (samcake) [inline interface/resources/qml/controls/CheckBox.qml]: Apparently a trailing end of file char ?

File: interface/src/scripting/WindowScriptingInterface.cpp [hunk 2]
```diff
@@ -45,13 +45,19 @@ QScriptValue WindowScriptingInterface::hasFocus() {
 }
 
 void WindowScriptingInterface::setFocus() {
-    auto window = Application::getInstance()->getWindow();
-    window->activateWindow();
-    window->setFocus();
+    // It's forbidden to call focus() from another thread.
+    Application::getInstance()->postLambdaEvent([] {
```
Comment (samcake) [inline interface/src/scripting/WindowScriptingInterface.cpp]: What is the post lambda event ?
what is it used for ?
who is calling for what purpose
Could the name be more explicit than saying it's a lambda ?

**Comment by samcake** [comment]
Aside from the 2 questions above it sounds good.
most of the logic is completely happening in qml at this point now
Cool!

File: interface/src/scripting/WindowScriptingInterface.cpp (hunk 2 above)
Comment (jherico) [inline interface/src/scripting/WindowScriptingInterface.cpp]: I thought the comment made it clear.  The WindowsScriptingInterface::setFocus() method can be called from any thread, but you can't call activateWindow() or focus() from another thread.  The `Application::postLambdaEvent()`  method is intended to be a mechanism to post something to be executed on the main thread without waiting for it to return.  

I could rename it something like 'executeOnMainThread' or 'queueForMainThread', but that seems kind of excessive.

Comment (samcake) [inline interface/src/scripting/WindowScriptingInterface.cpp]: OK now I get it
I would suggest you comment on the header application.h at the declaration of postLambdaEvent(
Just copy paste what you explained above ? I think it makes it clear,

**Pull request merged**